[DDP] Use compiled_autograd to trace DDP backward allreduce

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113209
* __->__ #110662

**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @Xia-Weiwen

fegin committed:
[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce

Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[WIP][DDP] Use compiled_autograd to trace DDP backward allreduce"


Use compiled_autograd to trace DDP backward allreduce.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

wconstab commented on test/distributed/_composable/test_replicate_with_compiler.py:
```diff
@@ -0,0 +1,184 @@
+
+        # Run multiple iterations so that we could test no_sync
+        for i in range(6):
+            # Settting a different random seed so that if the allreduces are not
```
typo

wconstab commented on test/distributed/_composable/test_replicate_with_compiler.py:
```diff
@@ -0,0 +1,184 @@
+                loss = model(input).sum()
+                loss.backward()
+
+            compiled_m = getattr(compiled_model, "_orig_mod", compiled_model)
```
confused about this line. if orig_mod doesn't exist, we just use compiled_model? In which cases does this happen?

fegin replied:
Good point. That seems to be some debug code. I changed it back.

wconstab commented on torch/_dynamo/skipfiles.py:
```diff
@@ -224,6 +227,7 @@ def _module_dir(m: types.ModuleType):
 if torch.distributed.is_available():
     MOD_INLINELIST.add("torch.distributed")
     MOD_INLINELIST.add("torch.distributed._functional_collectives")
+    MOD_INLINELIST.add("torch.distributed._composable.replicate")
```
@yanboliang are we making a new behavior for dynamo that should remove this list in the future? or are we just going to have to keep using this?

fegin replied:
This line is legit. The `LEGACY_MOD_INLINELIST` one is the one should be removed according to @yanboliang. But right now we have to add `LEGACY_MOD_INLINELIST` to make it work.

wconstab replied:
yea, that's what i meant. I was asking Yanbo if we are gonna have a new solution for LEGACY_... or if the fact we're still adding to it is a problem

wconstab commented on torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py:
```diff
@@ -57,24 +65,37 @@ def fp16_compress_hook(
+        if isinstance(bucket, tuple)
+        else bucket.buffer()
+    )
+    compressed_tensor = buffer.to(torch.bfloat16).div_(world_size)
```
(I see you have tests for bf16 and fp16. are the tests good enough if this is not correct and the passed?

fegin replied:
Thanks for catching the bug. The tests are legit. It compares the eager DDP with compiled DDP. Since both went to the same incorrect path (`buffer.to(torch.bfloat16)`), both got the same `incorrect` results. I think the issue is that why our DDP fp16 test didn't catch this error. I'll check this.

fegin replied:
I have to disable Inductor when enabling fp16. I don't think the issue is from tracing since disabling Inductor can make the test green. I'll follow up this issue.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -864,6 +866,44 @@ def __init__(
 
         self._lazy_init_ran = False
 
+        # Register the AccumulaGrad post hooks even if DDP is not compiled. This
```
typo Accumulate

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -864,6 +866,44 @@ def __init__(
+        self._accum_grad_hooks: List[RemovableHandle] = []
+        self._ddp_python_hook = torch._dynamo.config.ddp_python_hook
+        if self._ddp_python_hook:
+            torch._dynamo.config.optimize_ddp = False
```
i'm not sure we want to mutate the dynamo config from runtime.

instead, could we just error if this is set?  And expect users to either enable optimize_ddp or compile_ddp?

(its a bigger question of how we want to roll out this feature)

yf225 replied:
+1 I'll vote for having both `optimize_ddp` and `compile_ddp` config flags for now, and sunset both flags when compile DDP is ready for general rollout.

fegin replied:
Change the behavior to raise an exception.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -1386,8 +1429,18 @@ def _lazy_init(self):
 
     def _pre_forward(self, *inputs, **kwargs):
-        if not self._lazy_init_ran:
+        if self._ddp_python_hook and torch._utils.is_compiling():
```
wondering- is it cleaner to 
- ditch the `self._ddp_python_hook` flag,
- install the python hook inside _pre_forward under `if is_compiling()`
- not have to worry about removing the python hook under non-compiel mode

fegin replied:
* We cannot ditch `self._ddp_python_hook` because we could not register the python hook for users who still want to use `DDPOptimizer`. We could not distinguish `torch._dynamo.config.ddp_python_hook` from `torch._dynamo.config.optimize_ddp` by using only `is_compiling()`.

* I tried registering the hook in when both `is_compiing()` and `self._ddp_python_hook` are True. That resulted in recompilation in the second iteration (but none afterward). That's why I change to this design.

wconstab reviewed (commented):
overall LGTM! few nits and one possible bug. I can stamp after you reply

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

wconstab commented:
You ready for a review? Did you solve the issue about eager mixed with compiled ddp?

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

xmfan commented on torch/nn/parallel/distributed.py:
```diff
@@ -864,6 +866,49 @@ def __init__(
+                for hook, state in self._comm_hooks:
+                    hook(state, (param.grad, param))
+            else:
+                gradient = param.grad / self.process_group.size()
```
why do we divide the gradient here?

fegin replied:
DDP performs the average. We need to match the eager behavior.

fegin commented on torch/csrc/autograd/utils/lambda_post_hook.h:
```diff
@@ -9,24 +9,36 @@ namespace utils {
     return fn_(outputs, inputs);
   }
 
+  void compiled_args(CompiledNodeArgs& args) override {
+    TORCH_CHECK(
+        compiled_fn_ != nullptr,
+        "LambdaPostHook compiled_fn_ is not defined. Either disable "
+        "compiled_autograd or define compiled_fn_ for the LambdaPostHook.");
+    compiled_fn_(args);
+  }
```
This was added because the hope of tracing `LambdaPostHook` to enable eager DDP + compiled_autograd. But since we are not going to do that, we can just make it an no op as you suggested.

fegin commented:
@wconstab Yes, I added a new test to test the eager DDP + `compiled_autograd` with a private flag. The PR is ready for another review.

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

wconstab commented on torch/_dynamo/config.py:
```diff
@@ -233,6 +233,12 @@
+# and compiled_autograd(), the allreduce issued by DDP can be traced. This
+# flag is only used when DDP module is compiled.
+# This is an experimental feature.
+ddp_python_hook = False
```
also i found that the name 'ddp_python_hook' kept confusing me bc i expected that variable to be of type 'hook' rather than bool.

fegin replied:
Ideally, we should automatically enable the feature when compiled_autograd is enabled. But this is not doable with the current UX. Current DDP has to install the hooks when compiled_autograd is enabled and uninstall the hooks if compiled_autograd is disabled. Both need to happen before the backward() is called. With the current compiled_autograd(), we don't have much room to do without an explicit variable.

```
loss = model(input).sum()
with compiled_autograd(compile_fn):
    loss.backward()
```

An alternative is to always run the hooks and disable it if the `is_compiling()` is not True. But DDP has been used for some smaller models where CPU can be a bottleneck. It is unclear if we want to disrupt the existing DDP path.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -864,6 +866,49 @@ def __init__(
         self._lazy_init_ran = False
 
+        # Register the AccumulateGrad post hooks even if DDP is not compiled. This
+        # can avoid compiling the hooks twice. The Python hooks will be
```
why does adding the hooks avoid compiling hooks twice?

fegin replied:
If we dynamically detect if `compiled_autograd()` is enabled in the backward, and register the hook, it seems to trigger the compilation twice. Let me remove this comment since this cause more confusion than what it tries to explain.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -864,6 +866,49 @@ def __init__(
+            if not self.require_backward_grad_sync:
+                return
+
+            if self._comm_hooks:
```
so in the case we have _comm_hooks, those would point to a fp16 or bf16 hook, which would do compression as well as grad scaling.  if no comm hooks, we do our own grad scaling?

wconstab replied:
hmm, you added _comm_hooks as a new list in this PR, so it's not used by eager.

`dist._register_comm_hook(self.reducer, state, hook)` is used by eager.  Ideally i guess we could just peek at the list of hooks kept by reducer and use that directly.  but i guess that would require a new binding and that's annoying.  ok.

wconstab replied:
although, a bit of a deviation, but, since we're deviating a bit anyway, maybe we should just create another comm hook that does the grad scaling, and register that hook in the case we do not register f/bf16 hook.  Then inside compiled_accum_grad_hook we can just call _comm_hooks unconditionally and no need to have a conditional/gradscaler code.

fegin replied:
Yes, we deviate it from the cpp reducer comm_hook because the original reducer comm_hook also have to obtain some cpp reducer information. Given that we disable cpp reducer, it makes sense to have a separate comm_hook list.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -1355,8 +1400,11 @@ def _inside_ddp_forward(self):
 
     def _run_ddp_forward(self, *inputs, **kwargs):
-        with self._inside_ddp_forward():
+        if self._ddp_python_hook:
```
might be clearer if this were named 'compiled_ddp_enabled' or if you worry about confusing between current compiled graph-split ddp, 'compiled_ddp_v2' or something

fegin replied:
I changed it to `_use_python_reducer`. This should be more explicit and the confusion between the original ddp compilation and the new one.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -1385,9 +1433,22 @@ def _lazy_init(self):
         self._lazy_init_ran = True
 
+    def _should_disable_reducer(self) -> bool:
+        return torch._utils.is_compiling() or self._force_to_disable_reducer
```
add a comment here? - why the 'or' (which conditions are each half of the or true, and why isn't one condition sufficient)

wconstab replied:
oh.. i think this is just for testing?  maybe name it `_disable_reducer_for_testing`?

fegin replied:
This is not for testing. We have to disable the cpp reducer if we enable Python reducer. Change the name to `_should_disable_cpp_reducer`.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -1883,8 +1947,16 @@ def register_comm_hook(self, state: object, hook: Callable):
+            # by calling `dist.group.WORLD.size()`, which causes compilation
+            # errors. So we pre-decode the process group and pass it to the
+            # hook.
+            if state is None:
```
can you check if this is still true? @yifuwang has done some improvements to the funcol API and dynamo tracing support.  even if its not already fixed, wonder if we can just fix the dynamo part and not have to do this workaround.

yifuwang replied:
I'm not sure if this is fixed. But I can take a look

fegin replied:
I just tested, this is not fixed yet. I prefer to land this workaround first and investigate the root cause in a separate PR.

wconstab commented on torch/nn/parallel/distributed.py:
```diff
@@ -1385,9 +1433,22 @@ def _lazy_init(self):
+
     def _pre_forward(self, *inputs, **kwargs):
-        if not self._lazy_init_ran:
+        if self._ddp_python_hook and self._should_disable_reducer():
```
why is `self._ddp_python_hook` checked _outside_ of _should_disable_reducer, instead of putting the and condition _inside_ ? it seems like they are used 1:1

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

fegin commented:
@wconstab As we discussed offline, I converted `optimize_ddp` to `Union[bool, str]` and add an extra mode `python_reducer`. Although I also added `python_reducer_without_compiled_forward`, I didn't advertise this mode as we are not sure if this is a common use case.

wconstab commented on torch/_dynamo/config.py:
```diff
@@ -220,12 +220,19 @@
+# flag is 'python_reducer', DDP will disable C++ reducer and use Python reducer
+# to allow compiled_autograd to trace the communication and allow comm/compute
+# overlap without graph-breaks.
+optimize_ddp: Union[bool, str] = True
```
sigh.. 
config used to be an object.  so you could have setters/getters and some control over the format.  In my head I was envisioning that when the user sets the value (True/False) we'd parse that, convert it to the right enum value, and then getter would always produce enum.  Then in the code that checks the config we just rely on enum for less confusion.

since this is not the case anymore, i think the next best thing is to define a getter that we use from the code.  

config.get_optimize_ddp_mode() -> enum (or str if you prefer) which will do the type checking and normalize the true/false into one of the supported modes so the downstream code can always use the normalized way

fegin replied:
I prefer to convert the result to an enum but torch will complain having an enum in a config file. So I'll convert the boolean to str.

wconstab replied:
latest version lgtm.  Except, i was not sure why you need `m = sys.modules[__name__]`, instead of just accessing `optimize_ddp` ?

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]

wconstab commented on torch/_dynamo/config.py:
```diff
@@ -234,6 +234,28 @@
 # overlap without graph-breaks.
```
should update the comments here to talk about the modes.  I think you should put the new strings up front, and document something like this

'no_optimization' (False): (description..)
'ddp_optimizer' (True): (description..)
'python_reducer': (experimental, description...)
...
and then add a 'Note: bool values are mapped for backwards compatibility as shown above'

wconstab reviewed (approved):
nice work! but we still have to actually get it working e2e!

fegin committed:
Update on "[DDP] Use compiled_autograd to trace DDP backward allreduce"


**Summary**
The reducer of `DistributedDataParallel`  is implemented with C++ and it is not easy to trace the allreduce launched in the reducer. This PR modifies `DistributedDataParallel` to launch one allreduce per gradient when `compiled_autograd` is enabled. The changes allow us to use `compiled_autograd` to trace the allreduce and later be optimized (fused) in the Inductor.

**Key Logic**
1. If `ddp_python_hook` is True, we assume `compiled_autograd` is used. `DistributedDataParallel` registers `compiled_accum_grad_hook` for all parameters.
2. In the first forward() call, if `DistributedDataParallel` is not compiled, all  `compiled_accum_grad_hook` are deregistered. If `DistributedDataParallel` is compiled, all `compiled_accum_grad_hook` will be compiled by `compiled_autograd`.
3.  `compiled_accum_grad_hook` launches an allreduce to reduce the gradient of the parameter.

**Bucketing**
The compiled backward is slow because there is no bucketing for the allreduces. We rely on Inductor to bucket the allreduces.

The bucketing is done in a separate PR.

Differential Revision: [D49428482](https://our.internmc.facebook.com/intern/diff/D49428482/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng Xia-Weiwen

[ghstack-poisoned]
